Bugzilla/API/V1/Github.pm

```diff
@@ -0,0 +1,155 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::API::V1::Github;
+use 5.10.1;
+use Mojo::Base qw( Mojolicious::Controller );
+
+use Bugzilla::Attachment;
+use Bugzilla::Bug;
+use Bugzilla::Constants;
+use Bugzilla::Group;
+use Bugzilla::User;
+
+use Digest::SHA qw(hmac_sha256_hex);
+use Mojo::Util  qw(secure_compare);
+
+# Stolen from pylib/mozautomation/mozautomation/commitparser.py from
+# https://hg.mozilla.org/hgcustom/version-control-tools
+use constant BUG_RE => qr/
+  (
+    (?:
+      bug |
+      b= |
+      (?=\b\#?\d{5,}) |
+      ^(?=\d)
+    )
+    (?:\s*\#?)(\d+)(?=\b)
+  )/ix;
+
+sub setup_routes {
+  my ($class, $r) = @_;
+  $r->post('/github/pull_request')->to('V1::Github#pull_request');
+}
+
+sub pull_request {
+  my ($self) = @_;
+  Bugzilla->usage_mode(USAGE_MODE_MOJO_REST);
+
+  # Return early if not a pull_request event
+  my $event = $self->req->headers->header('X-GitHub-Event');
+  if (!$event || $event ne 'pull_request') {
+    return $self->code_error('github_pr_not_pull_request');
+  }
+
+  # Verify that signature is correct based on shared secret
+  if (!$self->verify_signature) {
+    return $self->code_error('github_pr_mismatch_signatures');
+  }
+
+  # Parse pull request title for bug ID
+  my $payload = $self->req->json;
+  if ( !$payload
+    || !$payload->{pull_request}
+    || !$payload->{pull_request}->{html_url}
+    || !$payload->{pull_request}->{title}
+    || !$payload->{pull_request}->{number})
+  {
+    return $self->code_error('github_pr_invalid_json');
+  }
+
+  my $html_url  = $payload->{pull_request}->{html_url};
+  my $title     = $payload->{pull_request}->{title};
+  my $pr_number = $payload->{pull_request}->{number};
+
+  $title =~ BUG_RE;
+  my $bug_id = $2;
+  my $bug    = Bugzilla::Bug->new($bug_id);
+  if ($bug->{error}) {
+    return $self->code_error('github_pr_bug_not_found');
+  }
+
+  # Check if bug already has this pull request attached
+  foreach my $attachment (@{$bug->attachments}) {
+    next if $attachment->contenttype ne 'text/x-github-pull-request';
+    if ($attachment->data eq $html_url) {
+      return $self->code_error('github_pr_attachment_exists');
+    }
+  }
+
+  # Create new attachment using pull request URL as attachment content
+  my $auto_user = Bugzilla::User->check({name => 'automation@bmo.tld'});
+  $auto_user->{groups}       = [Bugzilla::Group->get_all];
+  $auto_user->{bless_groups} = [Bugzilla::Group->get_all];
+  Bugzilla->set_user($auto_user);
+
+  my $timestamp = Bugzilla->dbh->selectrow_array("SELECT NOW()");
+
+  my $attachment = Bugzilla::Attachment->create({
+    bug         => $bug,
+    creation_ts => $timestamp,
+    data        => $html_url,
+    description => 'GitHub Pull Request',
+    filename    => "github-$pr_number-url.txt",
+    ispatch     => 0,
+    isprivate   => 0,
```

Comment:
Most of the time if someone creates a PR on GH it will be public so it is assumed the bug is public. But it doesn't have to be for this to work. The attachment will be created but if the user cannot see the bug they will also not be able to see the attachment or that it was made. We already do the same thing with regards to Phabricator in that we add the phabricator link attachment to the private bug without worrying if the Phabricator revision is private or not. Although with Phabricator we have other separate code that will make a revision private if the bug itself is private. We have no way to do that with Github PRs.